Create new rule/why-upgrade-to-latest-dotnet (#6943)
SSWConsulting/SSW.Rules.Content: rules/why-upgrade-to-latest-dotnet/rule.md

authors:
  - title: Yazhi Chen
    url: https://www.ssw.com.au/people/yazhi-chen

---
In the world of software development, staying up-to-date with technology is not just a trend; it's a necessity. As the digital landscape evolves, so do the tools and frameworks that developers rely on to build applications. One significant transition that has been taking place in recent years is the migration from the legacy .NET Framework to the latest .NET (currently .NET 7 as of November 2022 and .NET 8 will be released on November 2023) offerings. But why is this shift so crucial, and what benefits does it bring to the table?
<!--endintro-->

## The Evolution of .NET
To understand the significance of upgrading from .NET Framework to the latest .NET, let's first take a brief look at the evolution of the .NET platform.

### .NET Framework: The Legacy

The .NET Framework has been the backbone of Windows application development for nearly two decades. It has provided a robust and versatile environment for building Windows desktop applications, web applications, and services. However, as technology progressed, the limitations of .NET Framework became more apparent.

#### The Pain Point:

> * Compatibility: .NET Framework applications primarily run on Windows, limiting cross-platform deployment.

Reviewer: Change colons to dashes, it's easier to notice the split and aligns with the prefixes rules 

Take a look at https://www.ssw.com.au/rules/prefixes/

> * Compatibility: .NET Framework applications primarily run on Windows, limiting cross-platform deployment.

Reviewer: ```suggestion
* Compatibility: .NET Framework applications run on Windows, limiting cross-platform deployment.
```

Any .NET Framework not running on windows is likely mono (i.e.: Not .NET Framework) or some other flavor of windows. Unless you have a resource that suggests otherwise? Maybe I'm wrong?

Author: You are right. Double checked the Microsoft Doc.

> * Performance: Older versions of .NET Framework can be resource-intensive and less efficient.

Reviewer: Phrase this more positively?

```suggestion
* Performance: Each new version of .NET brings improved performance and resource utilization, allowing the same .NET code to perform better on the same hardware.
```

Author: This part is talking about the pain point of .NET framework, why we do not want to use it anymore.
And in the later part - ### Enter the Modern .NET(## Key Reasons to Upgrade), I am talking about the .NET, why do we want to use .NET to make the comparison.

Reviewer: Sure, you can say that .NET Framework is slow by saying newer .NET is fast 🙂. 

The way it's phrased at the moment, .NET 4.8.1 is faster and less resource intensive... because after that is .NET Core and then .NET, `older versions of .NET Framework` excludes the new versions of .NET Framework.

Author: Updated

> * Security: The framework reached a point where it was no longer actively developed, leaving applications vulnerable to security risks.

Reviewer: This line doesn't feel true

> Beginning with version 4.5.2 and later, .NET Framework is defined as a component of the Windows operating system (OS). Components receive the same support as their parent products, therefore, .NET Framework 4.5.2 and later follows the lifecycle policy of the underlying Windows OS on which it is installed.

https://learn.microsoft.com/en-us/lifecycle/faq/dotnet-framework

Author: Can I rephrase this to be "Some older versions of .NET Framework reached a point where it was no longer actively developed and supported, leaving applications vulnerable to security risks"? As said in the article, support for some older versions are ended in 2016 etc...

Reviewer: If you say that, why would I upgrade to .NET 8 instead of just upgrading to .NET Framework 4.8.1? It's still supported and my business case clearly doesn't need other features otherwise I would have moved already🙂

Author: Updated.

> * Modern development features: .NET Framework may not support some of the modern development features and patterns that are essential for staying competitive in today's tech landscape.

Reviewer: Phrase positively?

```suggestion
* Modern development features: Newer versions of .NET give us more opportunities to design products and services using many newer capabilities that are not backported to .NET Framework. This allows us to build, test and deploy faster, and more securely than ever before, keeping us competitive in today's tech landscape.
```

You might want to double-check the line to make sure I've not gone too far to the other side 😅

Author: This is the same, in this section, I want to focus on the pain point of the .NET Framework and why we do not want to stay in .NET Framework anymore.

Reviewer: It's just a suggestion 🙂. You can illustrate pain points through a positive for the alternative. 

Saying things like may not support modern development features makes it sound like you don't know if it does or not, I'd then make that be more assertive, as in, it doesn't support modern development features and patterns. 
Although the sentence is not completely accurate, as it depends on the business use case. 

This is where showing the alternative as the pain point comes across better

Author: Updated

### Enter the Modern .NET
Recognizing the need for change, Microsoft introduced a new direction for .NET with the release of .NET Core, which later evolved into .NET 5 and beyond. This modern .NET is designed to address the shortcomings of .NET Framework and meet the demands of contemporary software development.

## Key Reasons to Upgrade

> * Improved Performance and Efficiency

Reviewer: It looks like the intent here is headers, we should use h4's maybe, instead of bullets.

Also markdown like empty lines as separators so add in an extra 1 before and after the para to make sure it does the breaks as you'd expect

The latest .NET versions are optimized for better performance and resource utilization, resulting in faster and more efficient applications. This is crucial in today's fast-paced digital world, where users expect seamless and responsive software.
* Cross-Platform Compatibility

> The latest .NET is designed to be cross-platform. This means you can develop applications that run not only on Windows but also on macOS and Linux. This widens your reach and allows you to target a broader audience.

Reviewer: > This widens your reach and allows you to target a broader audience.

This sentence doesn't sit well with me because we're generally not building thick clients anymore. So our web apps allow us to target all customers already regardless of operating system.

It does, however, give us more flexibility in deployment, opening up the world of containerization, as well as using the same libraries in Web, WebAssembly (.NET Blazor), and native mobile apps using .NET MAUI

Author: updated

* Ongoing Support and Security

> The latest .NET versions receive regular updates, including security patches. This ongoing support ensures that your applications remain secure and up-to-date, protecting your data and your users.

Reviewer: Refer to 1 of my other points. This line is inaccurate as .NET Framework is supported full under windows, so if there were any security issues it would be patched as part of windows

https://learn.microsoft.com/en-us/lifecycle/products/microsoft-net-framework

https://learn.microsoft.com/en-us/lifecycle/faq/dotnet-framework#what-is-the-microsoft-lifecycle-policy-for--net-framework-

Author: Updated

* Modern Development Features

> The latest .NET provides access to modern development features, including support for C# 9 and above, enhanced tooling, and improved support for modern web development, such as Blazor for web applications.

Reviewer: C# 11 is out, and 12 comes out next month.

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version#defaults

Also, 90% of C# features can be used in .NET Framework as they are syntactical sugar

To try this out, add this to any .NET Framework project, I did this just today to make use of the cleaner using statements

```xml
<PropertyGroup>
   <LangVersion>11.0</LangVersion>
</PropertyGroup>
```

Author: Updated

> The latest .NET provides access to modern development features, including support for C# 9 and above, enhanced tooling, and improved support for modern web development, such as Blazor for web applications.

Reviewer: I don't think it's relevant saying .NET lets you use blazor, so you should upgrade... The support path unless you're adding blazor server to an existing app (I don't think that's recommended) is creating a new app regardless, so your old app can be .NET Framework while a new and shiny blazor app exists in the product

Author: Updated

> In conclusion, upgrading from .NET Framework to the latest .NET is not just a technological shift; it's a strategic move to future-proof your applications. It brings improved performance, cross-platform compatibility, ongoing support, and access to modern development features. By making this transition, you can stay competitive, secure, and ready to embrace the evolving landscape of software development.

Reviewer: I'd remove on going support as per some other comments (because it's there for .NET Framework)

Author: Updated

> In conclusion, upgrading from .NET Framework to the latest .NET is not just a technological shift; it's a strategic move to future-proof your applications. It brings improved performance, cross-platform compatibility, ongoing support, and access to modern development features. By making this transition, you can stay competitive, secure, and ready to embrace the evolving landscape of software development.

Author: Added a summary section

> 

Reviewer: End off with a link to .NET Framework support page

https://learn.microsoft.com/en-us/lifecycle/products/microsoft-net-framework

and .NET support page

https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core

    url: https://www.ssw.com.au/people/yazhi-chen

---
In the world of software development, staying up-to-date with technology is not just a trend; it's a necessity. As the digital landscape evolves, so do the tools and frameworks that developers rely on to build applications. One significant transition that has been taking place in recent years is the migration from the legacy .NET Framework to the latest .NET offerings. Read more about [.NET version lifecycle](https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core#lifecycle). But why is this shift so crucial, and what benefits does it bring to the table?
<!--endintro-->

## The Evolution of .NET

To understand the significance of upgrading from .NET Framework to the latest .NET, let's first take a brief look at the evolution of the .NET platform.

### .NET Framework: The Legacy

The .NET Framework has been the backbone of Windows application development for nearly two decades. It has provided a robust and versatile environment for building Windows desktop applications, web applications, and services. However, as technology progressed, the limitations of .NET Framework became more apparent.

#### The Pain Point:

> * Compatibility - .NET Framework applications run on Windows, limiting cross-platform deployment.

Reviewer: @YazhiChen we could maybe bold the prefixes so they stand out better on the page. At the moment it looks like a wall of text and I think this might help

Author: Hi @GordonBeeming, this is fixed in the PR #6959

* Security - Limited security support for older .NET Framework versions(4,4.5 and 4.5.1), and the support for NET Framework versions(4.5.2 and later) [follows the lifecycle policy of the underlying Windows OS on which it is installed](https://learn.microsoft.com/en-us/lifecycle/faq/dotnet-framework#what-is-the-microsoft-lifecycle-policy-for--net-framework-).
* Modern development features - Newer versions of .NET give us more opportunities to design products and services using many newer capabilities that are not backported to .NET Framework. This allows us to build, test and deploy faster, and more securely than ever before, keeping us competitive in today's tech landscape.

### Enter the Modern .NET

Recognizing the need for change, Microsoft introduced a new direction for .NET with the release of .NET Core, which later evolved into .NET 5 and beyond. This modern .NET is designed to address the shortcomings of .NET Framework and meet the demands of contemporary software development.

## Key Reasons to upgrade from .NET Framework to the Latest .NET

> ### Cross-Platform Compatibility

Reviewer: @YazhiChen these should be h4's, h3 doesn't style right on the pages

Author: Hi @GordonBeeming, this is fixed in the PR #6959
